### PR TITLE
[v1.27.0] chore: miscellaneous fixes

### PR DESCRIPTION
**Pull request by DavidMinarsch**
## Proposed changes

This PR fixes various small issues:
- `setup_class`/`teardown_class` misuse is fixed throughout
- meta classes are used to verify correct test case setup

## Fixes

https://github.com/valory-xyz/open-aea/issues/492

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes and CI passes too
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

-

### CODE REVIEW

File: aea/test_tools/click_testing.py [hunk 1]
```diff
@@ -147,12 +148,12 @@ def invoke(  # type: ignore
...
 
 
-class CliTest:
+class CliTest(ABC):
     """Test cli commands."""
 
-    t: Path
     cli_options: Sequence[str] = ()
 
+    _t: Path
```
Comment (DavidMinarsch) [inline aea/test_tools/click_testing.py:156]: This is private to the test case. Test case expects it to not be altered (see below)

File: aea/test_tools/test_contract.py [hunk 2]
```diff
@@ -39,11 +39,52 @@
...
+            )
+
+
+class BaseContractTestCase(ABC, metaclass=_MetaBaseContractTestCase):
     """A class to test a contract."""
 
-    path_to_contract: Path = Path(".")
-    ledger_identifier: str = ""
+    path_to_contract: Path
+    ledger_identifier: str
```
Comment (DavidMinarsch) [inline aea/test_tools/test_contract.py:87]: Setting default values here is undesirable: https://github.com/valory-xyz/open-aea/issues/492

Now we also validate via the meta class

File: aea/test_tools/test_contract.py [hunk 3]
```diff
@@ -66,19 +107,15 @@ def contract(self) -> Contract:
         return value
 
     @classmethod
-    def setup(cls, **kwargs: Any) -> None:
```
Comment (DavidMinarsch) [inline aea/test_tools/test_contract.py:69]: Setup gets called before each test method call. Defining it as a class method is confusing and leads to misuse.

File: aea/test_tools/test_skill.py [hunk 4]
```diff
@@ -43,10 +44,45 @@
...
+        """Check that required class attributes are set."""
+        if not hasattr(base_skill_test_case_cls, "path_to_skill"):
+            raise ValueError(f"'path_to_skill' not set on {base_skill_test_case_cls}")
+
+
+class BaseSkillTestCase(ABC, metaclass=_MetaBaseSkillTestCase):
     """A class to test a skill."""
 
-    path_to_skill: Path = Path(".")
+    path_to_skill: Path
```
Comment (DavidMinarsch) [inline aea/test_tools/test_skill.py:85]: see above

File: packages/fetchai/contracts/erc1155/tests/test_contract.py [hunk 5]
```diff
@@ -66,9 +66,9 @@ class TestERC1155ContractEthereum(BaseContractTestCase):
     path_to_contract = PACKAGE_DIR
 
     @classmethod
-    def setup(cls) -> None:
+    def setup_class(cls) -> None:
```
Comment (DavidMinarsch) [inline packages/fetchai/contracts/erc1155/tests/test_contract.py:69]: see above

**Review by 0xArdi — DISMISSED** [review]

**Review by 0xArdi — APPROVED** [review]

**Pull request merged**